server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -459,4 +466,152 @@ protected void onDelete() {
+
+        /**
+         * Total bytes to be uploaded to Remote Translog Store
+         */
+        long uploadBytes;
+
+        /**
+         * System nano time when the Remote Translog Store upload is started
+         */
+        long uploadStartTime;
+
+        /**
+         * System nano time when the Remote Translog Store upload is completed
+         */
+        long uploadEndTime;
+
+        /**
+         * System current time when the Remote Translog Store upload is completed
+         */
+        long uploadEndTimeMs;
+
+        public RemoteFsTranslogTransferListener(Releasable transferReleasable, Long generation, Long primaryTerm) {
+            this.transferReleasable = transferReleasable;
+            this.generation = generation;
+            this.primaryTerm = primaryTerm;
+        }
+
+        @Override
+        public void beforeUpload(TransferSnapshot transferSnapshot) throws IOException {
+            toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+            toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogFileSnapshots()));
+            toUpload.addAll(fileTransferTracker.exclusionFilter((transferSnapshot.getCheckpointFileSnapshots())));
+
+            if (toUpload.size() > 0) {
+                uploadBytes = RemoteStoreUtils.getTotalBytes(toUpload);
+                captureStatsBeforeUpload();
+                uploadStartTime = System.nanoTime();
+            }
+        }
+
+        @Override
+        public void onUploadComplete(TransferSnapshot transferSnapshot) throws IOException {
+            if (toUpload != null && toUpload.size() > 0) {
+                uploadEndTime = System.nanoTime();
+                uploadEndTimeMs = System.currentTimeMillis();
+                captureStatsOnUploadSuccess();
+            }
+
+            transferReleasable.close();
+            closeFilesIfNoPendingRetentionLocks();
+            maxRemoteTranslogGenerationUploaded = generation;
+            minRemoteGenReferenced = getMinFileGeneration();
+            logger.trace("uploaded translog for {} {} ", primaryTerm, generation);
+        }
+
+        @Override
+        public void onUploadFailed(TransferSnapshot transferSnapshot, Exception ex) throws IOException {
+            if (toUpload != null && toUpload.size() > 0) {
+                uploadEndTime = System.nanoTime();
+                captureStatsOnUploadFailure();
+            }
+
+            transferReleasable.close();
+            closeFilesIfNoPendingRetentionLocks();
+            if (ex instanceof IOException) {
+                throw (IOException) ex;
+            } else {
+                throw (RuntimeException) ex;
+            }
+        }
+
+        /**
+         * Adds relevant stats to the tracker when an upload is started
+         */
+        private void captureStatsBeforeUpload() {
+            remoteTranslogTransferTracker.addUploadsStarted(toUpload.size());
```

Comment:
for segments uploads, we are tracking uploads on entire refresh level. Should we track here the complete translog upload (which means tlog/ckp and metadata file upload) as one upload?